internal/mithril-metric/src/commons.rs

```diff
@@ -0,0 +1,494 @@
+use prometheus::{core::Collector, Counter, Gauge, Opts};
+use slog::{debug, Logger};
+
+use mithril_common::{entities::Epoch, StdResult};
+
+/// Type alias for a metric name.
+pub type MetricName = str;
+
+/// Type alias for a counter value.
+pub type CounterValue = u32;
+
+/// Create a MetricService.
+///
+/// To build the service you need to provide the structure name and a list of metrics.
+/// Each metrics is defined by an attribute name, a type, a metric name and a help message.
+/// When the metric name is not provided the attribute name will be used.
+/// We can specify a prefix that will be added to all metric names.
+///
+/// The attribute name will be used to create a getter method for the metric.
+///
+/// Crate that use this macro should have `paste` as dependency.
+///
+/// build_metrics_service!(
+///     MetricsService,
+///     counter_example: MetricCounter(
+///         "custom_counter_example_name",
+///         "Example of a counter metric"
+///     ),
+///     gauge_example: MetricGauge(
+///         "custom_gauge_example_name",
+///         "Example of a gauge metric"
+///     )
+/// );
+///
+/// build_metrics_service!(
+///     MetricsService,
+///     counter_example: MetricCounter("Example of a counter metric"),
```

Comment:
IMO we should always create a `Metric` with a given metric name to avoid side effects if e.g. we rename the field in the service.

Reply:
Removing this possibility, I also simplify build_metrics_service macro.